### ISSUE TITLE
Support only restore volume data of a backup

### ISSUE THREAD

**Issue by blackpiglet** [vmware-tanzu/velero#7345]
**Describe the problem/challenge you have**

Sometimes, it's possible to only restore data into volume without modifying the k8s resources.
Some issues required a similar function, #504 and #2598.
Suggest, at least, provide a workaround for this scenario.

**Describe the solution you'd like**

If a PodVolumeRestore is created based on the PodVolumeBackup information manually, the PodVolumeRestore is not processed by the Velero PVR controller. This is due to the PVR controller checking the PVR-requested pod whether has a restore-helper InitContainer. If not, the PVR is neglected.

Suggest adding a new field in the PVR to make the PVR controller skip the restore-helper InitContainer check logic. 
`AllowOnlyRestoreData` is the proposed parameter name. Its type is bool. The default value is false. This is an example:
``` yaml
apiVersion: velero.io/v1
kind: PodVolumeRestore
metadata:
  name: test-01-xxxxxxxxxxxxx-xxxxxx
  namespace: velero
spec:
  backupStorageLocation: default
  pod:
    kind: Pod
    name: hello-app-xxxxxxxxx-xxxxx
    namespace: restore
    uid: xxxxxxxx
  repoIdentifier: gs:jxun:/restic/test
  snapshotID: xxxxxxxx
  sourceNamespace: test
  uploaderSettings:
    WriteSparseFiles: "false"
  uploaderType: kopia
  volume: sdk-volume
  allowOnlyRestoreData: true
```

**Anything else you would like to add:**

**Environment:**

- Velero version (use `velero version`):
- Kubernetes version (use `kubectl version`):
- Kubernetes installer & version:
- Cloud provider or hardware configuration:
- OS (e.g. from `/etc/os-release`):

**Vote on this issue!**

This is an invitation to the Velero community to vote on issues, you can see the project's [top voted issues listed here](https://github.com/vmware-tanzu/velero/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).  
Use the "reaction smiley face" up to the right of this comment to vote.

- :+1: for "The project would be better with this feature added"
- :-1: for "This feature will not enhance the project in a meaningful way"

**Comment by blackpiglet** [comment]
As a PoC, already verified that, after bypassing the InitContainer check code, the manual created PVB worked as expected.

**Comment by stefanhenseler** [comment]
@blackpiglet, this is amazing! We need this functionality as well. Do you have a link to your PoC branch, would try it out and provide feedback as well. Is this a feature being worked on for 1.14?

**Comment by reasonerjt** [comment]
I think we need a design to elaborate the e2e flow and explain why this requires a change in the CRD.

**Comment by blackpiglet** [comment]
@stefanhenseler 
https://github.com/blackpiglet/velero/tree/7345_poc
I created a branch in my forked repository.
This is still a proof of concept of this issue. 
The change bypasses checking the restore-help InitContainer for PodVolumeRestore pointing pod.
The user needs to create the PodVolumeRestore with the correct content, and the target pod and PVC should exist before creating the PVR.

**Comment by qmaraval-csgroup** [comment]
Hello, i have an issue similar to this idea.

The context: 
I use velero to perform backups of PVC using FSB.
I only want to backup the content of my PVC, from what i read i have to backup at least : 
- The pod (to allow the initcontainer to be scheduled) where the PVC is mounted on
- The PV 
- The PVC 


Everything works well for backup but the restoration is a little bit more adventurous... 
I want to restore PVC directly on the backup namespace, however as the pod comes from a statefulset or a deployment, the restoration pod can not be scheduled, so the init container never run and no restore is performed.
Do you think it's possible through restore-resource-modifiers to generate a fake pod modifying the initial one just to allow init container to run (bypassing STS and deployment own policy) ?

For now, the only way i have to make it possible is to backup all resources from Namespace A, restore them in another empty namespace B, once everything is restored, i change the PV reclaim policy, delete everything and manually reattach the pv from B to the PVC on  namespace A ... Which works but is far from optimal ...

Thanks,

**Comment by blackpiglet** [comment]
@qmaraval-csgroup 
Thanks for sharing your thoughts.
Could you share more information about why you only want to back up and restore the data in volume?

I know one possible scenario is that the k8s workload is governed by GitOps tools. Is this the same as your scenario?

**Comment by qmaraval-csgroup** [comment]
@blackpiglet 
Thanks for your answer. 
Indeed the reason is that we kind of use GitOps tool, it is not really a gitops tool but a specific CI/CD.

**Comment by blackpiglet** [comment]
@qmaraval-csgroup
Velero cannot use the resource modifier to remove the InitContainer from the PodVolumeRestore-generated pod, because, from the Velero perspective, the pod is already restored, Velero doesn't have a phase in restore to modify the already restored resource.

I see. Using the k8s in the DevOps pipeline to deploy the new version application into the production or staging environment is a typical scenario. Usually, these DevOps tools will guarantee the workload's state and replica. If something goes wrong, it will send alerts to the contact.

Thanks for your feedback. Just from curiosity, does this environment often perform backup and restore operations?

**Comment by qmaraval-csgroup** [comment]
@blackpiglet 
Thanks for your answer.
The backup are performed on a daily basis. 
The restore are rarer maybe once or twice a month according to users mistakes.

**Comment by DreamingRaven** [comment]
Adding my two cents for why I would desire this feature.

I have a production cluster. I also have a staging cluster. These two clusters are divergent in that environment variables, secrets, configmaps, etc will be different due to their two very different environments E.G one is hosted kubernetes and another is Talos and have different ingress hosts like app.org.example vs app.staging.org.example in staging. This means it is undesirable for me to attempt to restore manifests from one cluster to another cluster, since they will be different. Both clusters are also GitOps clusters so they have no need of externally provided manifests. They both have CSI snapshot support.

I want this data from production in staging to enable me to test before new changes enter production using real data. Let's say a series of databases like MongoDB or data stores like Minio that need the data to be that of production to be able to fully test new changes. Thus, the only thing I require is the volumes themselves to be moved.

Hopefully that is a clear potential use-case for this restore.

**Comment by blackpiglet** [comment]
unstale

**Comment by simenhaga** [comment]
> Hello, i have an issue similar to this idea.
> 
> The context: I use velero to perform backups of PVC using FSB. I only want to backup the content of my PVC, from what i read i have to backup at least :
> 
> * The pod (to allow the initcontainer to be scheduled) where the PVC is mounted on
> * The PV
> * The PVC
> 
> Everything works well for backup but the restoration is a little bit more adventurous... I want to restore PVC directly on the backup namespace, however as the pod comes from a statefulset or a deployment, the restoration pod can not be scheduled, so the init container never run and no restore is performed. Do you think it's possible through restore-resource-modifiers to generate a fake pod modifying the initial one just to allow init container to run (bypassing STS and deployment own policy) ?
> 
> For now, the only way i have to make it possible is to backup all resources from Namespace A, restore them in another empty namespace B, once everything is restored, i change the PV reclaim policy, delete everything and manually reattach the pv from B to the PVC on namespace A ... Which works but is far from optimal ...
> 
> Thanks,

We experience exactly the same problem as [qmaraval-csgroup](https://github.com/qmaraval-csgroup), but we use ArgoCD for GitOps. We need to be able to recover all data in case of  a disaster recovery situation. When restoring from cluster-1 to cluster-2 we are only able to get the data from the PVC if the namespace on the restore side is totally empty, but as we are governed by gitops this will not be the case in an emergency. 

Are you still working on solving the problem and do you know when we can assume it is solved?

Any updates would be appreciated.

**Comment by blackpiglet** [comment]
Yes, but this issue is not planned in the v1.15 release, so no ETA yet.

**Comment by 2fst4u** [comment]
> Moving this out of v1.14 milestone as we are seeing different understanding about the feature.
> More discussion is needed before we have consensus about the scope.

Is there anything further that can be contributed to aid in understanding the scope and requirements of this request?

I'm in the same boat as everyone else, my manifests are all in gitops so I really have no need to backup all my objects, but only the data contents of my PVs. In a disaster I'd just need to bootstrap my cluster with flux again and I'd have everything all set up, with the exception of the persistent data. Velero restore would be redundant for my kubernetes objects.

Not asking for an ETA at all just wondering if there is anything further we can help with to aid in this.

**Comment by kaovilai** [comment]
Requirements are pretty clear atm. A valid use case is there. We just need bandwidth to figure out design and implementation, which anyone able to is welcome to PR some draft ideas or as a comment here. Some pitfalls, cautionary tales also helps if you know any.

**Comment by the-veloper** [comment]
How would the restore happen PVCs with the `ReadWriteOnce` access mode? There are some CSI drivers where `ReadWriteMany` is not supported and velero would not be able to schedule the restore pod, if the PVC is already mounted to the actual workload that uses the PVC.

**Comment by pschichtel** [comment]
RWO PVCs can be mounted by multiple pods on the same node, so a Job would work. RWOP prevents that, but that could be worked around by injecting ephemeral containers into the pod that mounts it.

I wonder though: is it actually wise to have the application running while restoring it's state? Replacing files while the application might already be working on them can cause all sorts of issues. I think stopping the pod would be better, but then you are fighting controllers starting a new pod. Scaling the deployment/statefulset might work to some degree, but then you'll be fighting CD tools trying to self-heal their deployment. An admission controller could be setup that prevents recreating the pod while the restore is in progress. Alternatively, some form of integration with CD tools might be imaginable, so there would be a way for velero to tell e.g. Argocd not to self-heal the application during the restore, maybe through annotations/labels. 

Depending on how the backup was made, additional steps might be necessary. For a example a backup of a postgres container might have a pre-backup command to run pg_dump, so it would also need a post-restore command to pg_restore the dump.

File permissions might be another issue, depending on how the backup was done, exact file permissions might not be retained. Not sure if that is the case in velero, but worth keeping in mind. Applications would need to be able to recover from wrong permissions or velero needs to somehow figure out the right permissions and restore them automatically.

For me the use-case is the same: my resources are mostly managed by gitops tools, so backups are not necessary with a few exceptions (operator generated secrets for example), so I'm mostly concerned with restoring PVCs. In the past I went the manual route by using restic to restore the files from the backup repository and copying that into the PVCs. So for me a restore of a velero backup means: restoring a small subset of resources and restoring pvc content.

**Comment by WRKT** [comment]
Hello!

Is this issue still onboarding ? 
Would appreciate to also have this feature because some of our storageClass doesn't support csi snapshot.

Thanks again to the velero team!

**Comment by kaovilai** [comment]
yes

**Comment by blackpiglet** [comment]
IMO, the in-place data-only restore is still prioritized.

Some modifications from the original design:
* Need to consider whether that's reasonable to apply the ResourcePolicy's VolumePolicy to restore. It would be easier to filter out the volumes that need a data-only restore. To make things clear, we can just support the action `InPlaceDataOnlyRestore`.
* Need to introduce a new hook. The hook is supposed to run on the pod with volume requiring the data-only restore. That will give a chance to cleanse the pod's volume w/r operations before restoring.
* The in-place data-only restore should only care about the PVC. The PV and pod can be different from the backup. In this way, it can support more scenarios, such as GitOps.
* Basically, the logic will be added in the function `restoreItem`. The RIA may not be useful here, because the RIA cannot know much runtime information about k8s resources.

**Comment by reasonerjt** [comment]
With VGDP it would be possible to restore data to the volume not only handled via fs-backup.  Updated the issue.

**Comment by reasonerjt** [comment]
In an internal discussion, there are disagreements on whether we should provide a feature that we know can only partially cover the case.  Let's discuss more in community meeting @shubham-pampattiwar

**Comment by blackpiglet** [comment]
After a recent discussion, we still didn't agree on whether this should be implemented and how it should work.
As a result, remove the 1.18-candidate label.
